src/ott/neural/flows/genot.py

```diff
@@ -0,0 +1,490 @@
+      rng: Optional[jax.Array] = None,
+      forward: bool = True,
+      diffeqsolve_kwargs: Dict[str, Any] = types.MappingProxyType({}),
+  ) -> Union[jnp.array, diffrax.Solution, Optional[jnp.ndarray]]:
+    """Transport data with the learnt plan.
+
+    This method pushes-forward the `source` to its conditional distribution by
+      solving the neural ODE parameterized by the
+      :attr:`~ott.neural.flows.genot.velocity_field`
+
+    Args:
+      source: Data to transport.
+      condition: Condition of the input data.
+      rng: random seed for sampling from the latent distribution.
+      forward: If `True` integrates forward, otherwise backwards.
+      diffeqsolve_kwargs: Keyword arguments for the ODE solver.
+
+    Returns:
+      The push-forward or pull-back distribution defined by the learnt
+      transport plan.
+
+    """
+    rng = utils.default_prng_key(rng)
+    if not forward:
+      raise NotImplementedError
+    diffeqsolve_kwargs = dict(diffeqsolve_kwargs)
+    assert len(source) == len(condition) if condition is not None else True
+
+    latent_batch = self.latent_noise_fn(rng, shape=(len(source),))
+    cond_input = source if condition is None else jnp.concatenate([
+        source, condition
+    ],
+                                                                  axis=-1)
+    t0, t1 = (0.0, 1.0)
+
+    @jax.jit
+    def solve_ode(input: jnp.ndarray, cond: jnp.ndarray):
+      return diffrax.diffeqsolve(
+          diffrax.ODETerm(
+              lambda t, x, args: self.state_velocity_field.
+              apply_fn({"params": self.state_velocity_field.params},
+                       t=t,
+                       x=x,
+                       condition=cond)
+          ),
+          diffeqsolve_kwargs.pop("solver", diffrax.Tsit5()),
+          t0=t0,
+          t1=t1,
+          dt0=diffeqsolve_kwargs.pop("dt0", None),
+          y0=input,
+          stepsize_controller=diffeqsolve_kwargs.pop(
+              "stepsize_controller",
+              diffrax.PIDController(rtol=1e-5, atol=1e-5)
+          ),
+          **diffeqsolve_kwargs,
+      ).ys[0]
+
+    return jax.vmap(solve_ode)(latent_batch, cond_input)
+
+  def _valid_step(self, valid_loader, iter):
+    """TODO."""
+    next(valid_loader)
+
+  @property
+  def learn_rescaling(self) -> bool:
+    """Whether to learn at least one rescaling factor."""
+    return self.rescaling_a is not None or self.rescaling_b is not None
+
+  def save(self, path: str):
+    """Save the model.
+
+    Args:
+      path: Where to save the model to.
+    """
+    raise NotImplementedError
+
+  def load(self, path: str) -> "GENOT":
+    """Load a model.
+
+    Args:
+      path: Where to load the model from.
+
+    Returns:
+      An instance of :class:`ott.neural.solvers.OTFlowMatching`.
+    """
+    raise NotImplementedError
+
+  @property
+  def training_logs(self) -> Dict[str, Any]:
+    """Logs of the training."""
+    raise NotImplementedError
+
+  def sample_noise(
```

Comment:
Unless you disagree or I'm missing something, I'd still put this into the `BaseFlow` class.